### PR TITLE
[Core][ContactStructuralMechanicsApplication] Update OBB. Addin…

### PR DESCRIPTION
**Pull request by loumalouomega**
Currently the OBB can be build providing a geometry. From this geometry it takes the AABB, the lowest and highest point, and builds and axis and defines an orthonormal base. This PR adds an option (not default), which allows to build from the normal (in case of line or surface). Taking this normal defines a base, rotates the nodes and gets the node at the maximum distance from the geometry center. Combining the tangent base defines a vector in the direction to the most far point, the third axis is taken as the cross product of the normal and this vector. This new method defines a less arbitrary bounding box than the previous one: Difference:

- Old method:
imagen

- New method:
imagen

The old one may look more regular, but just takes an arbitrary direction as all the nodes are in the same plane. The new one defines a larger OBB and oriented to the most far node from the center, which is less arbitrary.

This option is not the default, until I do more research

### CODE REVIEW

**Comment by loumalouomega** [comment]
BTW, it makes no difference on 2D

File: kratos/sources/oriented_bounding_box.cpp [hunk 1]
```diff
@@ -82,22 +84,57 @@ OrientedBoundingBox<2>::OrientedBoundingBox(
...
-    rGeometry.BoundingBox(geometry_low_node, geometry_high_node);
+    if (BuildFromBoundingBox || rGeometry.WorkingSpaceDimension() == rGeometry.LocalSpaceDimension()) {
+        // Check the intersection of each face of the object bounding box against the intersecting object bounding box
+        NodeType geometry_low_node, geometry_high_node;
+        rGeometry.BoundingBox(geometry_low_node, geometry_high_node);
+
+        // Creating OBB
+        noalias(mOrientationVectors[0]) = geometry_high_node - geometry_low_node;
+        const double norm_mOrientationVectors = norm_2(mOrientationVectors[0]);
+        if (norm_mOrientationVectors > std::numeric_limits<double>::epsilon())
```
Comment (rubenzorrilla) [inline kratos/sources/oriented_bounding_box.cpp]: I'm wondering if the `std::numeric_limits<double>::epsilon()` could lead to valid but ill-conditioned cases.

File: kratos/sources/oriented_bounding_box.cpp [hunk 2]
```diff
@@ -82,22 +84,57 @@ OrientedBoundingBox<2>::OrientedBoundingBox(
...
+            mOrientationVectors[0] /= norm_mOrientationVectors;
+        else
+            KRATOS_ERROR << "Zero norm on OrientedBoundingBox direction" << std::endl;
```
Comment (rubenzorrilla) [inline kratos/sources/oriented_bounding_box.cpp]:
```suggestion
            KRATOS_ERROR << "Zero norm in OrientedBoundingBox direction" << std::endl;
```

File: kratos/tests/cpp_tests/processes/test_find_intersected_geometrical_objects_process.cpp [hunk 3]
```diff
@@ -28,17 +29,17 @@
 namespace Kratos {
     namespace Testing {
 
-        typedef Node<3> NodeType; // TODO: Replace in the rest of the test
```
Comment (rubenzorrilla) [inline kratos/tests/cpp_tests/processes/test_find_intersected_geometrical_objects_process.cpp:31]: :+1:

File: kratos/includes/oriented_bounding_box.h [hunk 4]
```diff
@@ -134,7 +134,8 @@ class KRATOS_API(KRATOS_CORE) OrientedBoundingBox
      */
     OrientedBoundingBox(
         const GeometryType& rGeometry,
-        const double BoundingBoxFactor
+        const double BoundingBoxFactor,
+        const bool BuildFromBoundingBox = true
```
Comment (rubenzorrilla) [inline kratos/includes/oriented_bounding_box.h:138]: According to the description of the PR, shouldn't this be `false` by default?

File: kratos/sources/oriented_bounding_box.cpp (hunk 1 above)
Comment (loumalouomega) [reply kratos/sources/oriented_bounding_box.cpp]: This is already working...

File: kratos/includes/oriented_bounding_box.h (hunk 4 above)
Comment (loumalouomega) [reply kratos/includes/oriented_bounding_box.h:138]: No, true is the the current behavior, false is the new one

File: kratos/sources/oriented_bounding_box.cpp (hunk 1 above)
Comment (loumalouomega) [reply kratos/sources/oriented_bounding_box.cpp]: I willl define a static value, wo if we need to update only is needed once

**Review by AlejandroCornejo — APPROVED** [review]
I understand that this is not the default until you check its robustness

**Pull request merged**